CommonTools/RecoAlgos/src/PrimaryVertexAssignment.cc

```diff
@@ -4,50 +4,102 @@
 #include "DataFormats/Math/interface/deltaR.h"
 #include "TrackingTools/IPTools/interface/IPTools.h"
-
+#include "FWCore/Utilities/interface/isFinite.h"
 
 
 std::pair<int,PrimaryVertexAssignment::Quality>
 PrimaryVertexAssignment::chargedHadronVertex( const reco::VertexCollection& vertices,
                                    const reco::TrackRef& trackRef,
                                    const reco::Track* track,
+                                   float time, 
+                                   float timeReso, // <0 if timing not available for this object
                                    const edm::View<reco::Candidate>& jets,
                                    const TransientTrackBuilder& builder) const {
 
-  int iVertex = -1;
-  size_t index=0;
   typedef reco::VertexCollection::const_iterator IV;
   typedef reco::Vertex::trackRef_iterator IT;
+
+  int iVertex = -1;
+  size_t index=0;
   float bestweight=0;
   for( auto const & vtx : vertices) {
       float w = vtx.trackWeight(trackRef);
       if (w > bestweight){
-          bestweight=w;
-          iVertex=index;
-        }        
+        bestweight=w;
+        iVertex=index;
+      }        
       index++;
   }
+  
+  bool useTime = useTiming_;
+  if (edm::isNotFinite(time) || timeReso<1e-6) {
+    useTime = false;
+    time = 0.;
+    timeReso = -1.;
+  }
+  
+  if (preferHighRanked_) {
+    for(IV iv=vertices.begin(); iv!=vertices.end(); ++iv) {
+        int ivtx = iv - vertices.begin();
+        if (iVertex == ivtx) return std::pair<int,PrimaryVertexAssignment::Quality>(ivtx,PrimaryVertexAssignment::UsedInFit);
+        
+        double dz = std::abs(track->dz(iv->position()));
+        double dt = std::abs(time-iv->t());
+        
+        bool useTimeVtx = useTime && iv->tError()>0.;
+        
+        if ((dz < maxDzForPrimaryAssignment_ or dz/track->dzError() < maxDzSigForPrimaryAssignment_ ) and (!useTimeVtx or dt/timeReso < maxDtSigForPrimaryAssignment_)) {
+          return std::pair<int,PrimaryVertexAssignment::Quality>(ivtx,PrimaryVertexAssignment::PrimaryDz);
+        }               
+    }
+  }
 
-
+    
   if(iVertex >= 0 ) return std::pair<int,PrimaryVertexAssignment::Quality>(iVertex,PrimaryVertexAssignment::UsedInFit);
-
-    double dzmin = 1e99;
-    int vtxIdMinDz = -1;
-    for(IV iv=vertices.begin(); iv!=vertices.end(); ++iv) {
-      double dz = std::abs(track->dz(iv->position()));
-      if(dz<dzmin) {
-        dzmin = dz;
-        vtxIdMinDz = iv-vertices.begin();
-      }
-   }
+    
+  double distmin = std::numeric_limits<double>::max();
+  double dzmin = std::numeric_limits<double>::max();
+  double dtmin = std::numeric_limits<double>::max();
+  int vtxIdMinDist = -1;
+  for(IV iv=vertices.begin(); iv!=vertices.end(); ++iv) {
+    double dz = std::abs(track->dz(iv->position()));
+    double dt = std::abs(time-iv->t());
+    
+    double dzsig = dz/track->dzError();
+    double dist = dzsig*dzsig;
```

Comment:
why is the vertex uncertainty not a part of the computation?
Is it because of more frequent possible assignment to poorly reconstructed vertices?

Reply:
Yes, that is the reason.

Actually I do recall now that the vertex uncertainty IS used in once place later in the file.  Let me double check what the reason was.